atl-checker/src/algorithms/certain_zero/search_strategy/linear_optimize.rs

```diff
@@ -139,322 +124,195 @@ impl LinearOptimizeSearch {
+                let until_symbol_range_map = self.map_phi_to_constraints(until);
 
-                RangedPhi::Or(
+                LinearConstrainedPhi::Or(
                     Box::from(pre_symbol_range_map),
                     Box::from(until_symbol_range_map),
                 )
             }
-            Phi::DespiteEventually { formula, .. } => self.map_phi_to_ranges(formula),
-            Phi::EnforceEventually { formula, .. } => self.map_phi_to_ranges(formula),
-            Phi::DespiteInvariant { formula, .. } => self.map_phi_to_ranges(formula),
-            Phi::EnforceInvariant { formula, .. } => self.map_phi_to_ranges(formula),
+            Phi::DespiteEventually { formula, .. } => self.map_phi_to_constraints(formula),
+            Phi::EnforceEventually { formula, .. } => self.map_phi_to_constraints(formula),
+            Phi::DespiteInvariant { formula, .. } => self.map_phi_to_constraints(formula),
+            Phi::EnforceInvariant { formula, .. } => self.map_phi_to_constraints(formula),
         }
     }
 
-    /// Goes through the RangedPhi and finds how close we are to acceptance border in this state.
-    fn visit_ranged_phi(&self, ranged_phi: &RangedPhi, state: &State) -> Option<i32> {
-        match ranged_phi {
-            // If we need to satisfy either of the formulas, just return the lowest distance found between the two
-            // If one the the sides is None, the other is returned (given that the other is Some)
-            // This makes sure that "x < 5 || false" will not return None, but the dist in lhs
-            RangedPhi::Or(lhs, rhs) => {
-                if let Some(lhs_distance) = self.visit_ranged_phi(lhs, state) {
-                    return if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
-                        Some(Ord::min(lhs_distance, rhs_distance))
-                    } else {
-                        Some(lhs_distance)
-                    };
-                } else if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
-                    return Some(rhs_distance);
-                }
-                None
+    /// Takes an expression and maps it to a LinearConstrainedPhi
+    fn map_expr_to_constraints(&self, expr: &Expr) -> LinearConstrainedPhi {
+        match &expr.kind {
+            ExprKind::UnaryOp(UnaryOpKind::Not, sub_expr) => {
+                return self.map_expr_to_constraints(sub_expr);
             }
-            // If we need to satisfy both of the formulas, just return the largest distance found between the two
-            RangedPhi::And(lhs, rhs) => {
-                if let Some(lhs_distance) = self.visit_ranged_phi(lhs, state) {
-                    if let Some(rhs_distance) = self.visit_ranged_phi(rhs, state) {
-                        return Some(Ord::max(lhs_distance, rhs_distance));
+            ExprKind::BinaryOp(operator, lhs, rhs) => {
+                match operator {
```

Comment:
We should be able to add more cases here, implication, ternary and so on can be converted to a disjunctive/conjunctive normal form

Reply:
Feel free to add those changes. I am working on floats.

Reply:
I'll look into it

Reply:
#163